tornado/test/wsgi_test.py

```diff
@@ -1,20 +1,116 @@
+import asyncio
+import concurrent.futures
+import threading
+
 from wsgiref.validate import validator
 
-from tornado.testing import AsyncHTTPTestCase
+from tornado.routing import RuleRouter
+from tornado.testing import AsyncHTTPTestCase, gen_test
 from tornado.wsgi import WSGIContainer
 
 
-class WSGIContainerTest(AsyncHTTPTestCase):
+class WSGIAppMixin:
     # TODO: Now that WSGIAdapter is gone, this is a pretty weak test.
-    def wsgi_app(self, environ, start_response):
+    def get_executor(self):
+        raise NotImplementedError()
+
+    def get_app(self):
+        executor = self.get_executor()
+        # The barrier test in DummyExecutorTest will always wait the full
+        # value of this timeout, so we don't want it to be too high.
+        self.barrier = threading.Barrier(2, timeout=0.3)
```

Comment:
Great solution for a test case! I did something much more brittle with sleeps elsewhere.